### PR TITLE
[WIP] Screenshots and screen recordings of tests

### PR DESCRIPTION
**Pull request by donataswix**

### CODE REVIEW

**Comment by rotemmiz** [comment]
This PR relates to https://github.com/wix/detox/issues/171

File: detox/src/devices/EmulatorDriver.js [hunk 1]
```diff
@@ -81,6 +82,66 @@ class EmulatorDriver extends AndroidDriver {
...
+      } catch (e) {
+        recording = false;
+      }
+    }
+
+    this._video = promise.childProcess;
+
+    function spawnRecording(width, height) {
+      return adb.spawn(deviceId, [
+        'shell', 'screenrecord', '--size', width + 'x' + height, '--verbose', '/sdcard/recording.mp4'
```
[the remark this answers is not in the document]
Comment (donataswix) [reply detox/src/devices/EmulatorDriver.js]: It was left over from my tests, will be removed.

File: detox/src/devices/EmulatorDriver.js [hunk 2]
```diff
@@ -81,6 +82,48 @@ class EmulatorDriver extends AndroidDriver {
...
+        promise = spawnRecording(width >>= 1, height >>= 1);
+        promise.catch(handleRecordingTermination);
+      }
+    }
+  }
+
+  async stopVideo(deviceId) {
+    if (this._video) {
+      this._video.kill(2);
+      await this.adb.adbCmd(deviceId, 'shell sleep 1');
```
Comment (rotemmiz) [inline detox/src/devices/EmulatorDriver.js]: I removed this line and saw the corrupted videos you talked about.
The thing is, this slows down the tests a lot! 
I am trying to think of a way to mitigate the penalty:
1. Let's create each file on the device with a different name.
2. when test is done, create  a copy+cleanup task
3. add this task to a queue and let this queue run independently of the test lifecycle.

WDYT ?

**Comment by Cederman** [comment]
Awesome job @donataswix! Question - I looked at the PR but I can't see the flexibility to give the screenshot (or video) a name.

As an example, If I want to document a user journey and have screenshots taken at different screens, I might want to name them "login", "profile", "settings" or something similar.

Could you add that to the PR?

**Comment by rotemmiz** [comment]
All artifacts are being picked up by ArtifactsCopier, which then renames them with the same name as the test description.

**Comment by Cederman** [comment]
@rotemmiz I see several issues with that, it will conflict (overwrite) or have "test description 1|2|3.png" if you would do several screenshot calls in the same test (assuming you can call it imperatively in the test). It also dictates the name itself which I don't see the need for, defaulting to the test description could work (although it might lead to long file names).

However, providing that functionality can come in a later PR but it would be nice to add it in this one. I can see it add value to several use cases.

**Comment by donataswix** [comment]
@Cederman this should work:
```javascript
it('...', () => {
  // do something
  await device._takeScreenshot('between-things');
  // do another thing
});
```
As you can tell by the `_` prefix, I didn't think of it as public API. It won't hurt to change it, I guess.

**Comment by haswalt** [comment]
Hello, any more progress on this? We've just picked up Detox and think it's great but we need the screenshot functionality to do visual regression and user flow testing.

**Comment by rotemmiz** [comment]
I hope we will merge this by the end of the week.

**Comment by haswalt** [comment]
any chance on this getting merged soon? I know you guys are pretty busy, happy to help out fixing the conflicts if needed.

**Comment by skovhus** [comment]
Do you guys need any help with this? Would be really nice to have.

**Comment by rotemmiz** [comment]
This is being actively worked on by @noomorph.
Updates can be found [here](https://github.com/wix/detox/issues/171)

@noomorph, can you update regarding the current state ? Is there a way to delegate/split some of the work ?

**Comment by noomorph** [comment]
@rotemmiz , sure! see my response in https://github.com/wix/detox/issues/171

**Comment by rotemmiz** [comment]
Closing this in favor of #734

**Pull request closed without merging**